[doc] Some clarifification and elaboration for ABI checking feature.

130s opened the pull request:
Supplement for https://github.com/ros-industrial/industrial_ci/pull/199

Something wasn't clear to me that is left as "TODO". If you can clarify I can update this PR.

mathias-luedtke committed:
implemented ABI check based on abi-compliance-checker

130s committed:
[doc] Some clarifification and elaboration for ABI checking feature.

mathias-luedtke commented on doc/index.rst:
```diff
@@ -150,10 +150,10 @@ Optional environment variables
-* `ABICHECK_FALLBACK` (default: not set): URL of archive to be used if no merge parent could be detected,
-* `ABICHECK_URL` (default: not set): URL of archive to check against. If empty, merge parent is used or `ABICHECK_FALLBACK` for regular commits.
-* `ABICHECK_VERSION` (default: not set): Readble version name. It will be read from the URL if possible.
+* `ABICHECK` (default: false): If `true`, run a binary compatibility check with `ABICC <https://github.com/lvc/abi-compliance-checker>`_ against the current branch (or the targeted branch in the pull/merge request), then exit (see `below <#abi-checks>`_ for details).
```
It is not checked against branches, but either against an archive or the left-hand merge parent (= target branch of PRs).
This way it works for any Git repo and  even for tests ABI checks in forks.
What do you mean by "current branch"?

mathias-luedtke commented:
Hmm, the docs really need some clarifications!
The idea to give some examples is great, but your examples are mixed up..
(our `.travis.yml` does not contain examples of the regular usage!)
The `_EXTERNAL_REPO` is just an internal feature (in `_wrap.sh`): it clones the external repo version and runs `industrial_ci` for it. It is not releated to `ABICHECK` at all.

mathias-luedtke commented on doc/index.rst:
````diff
@@ -251,6 +251,29 @@ So either `ABICHECK_URL` or `ABICHECK_FALLBACK` must be set.
+Case-1: Check ABI compatibility between current and the specific version.
+
+```
+- ROS_DISTRO=indigo _EXTERNAL_REPO='ros/actionlib#38ce66e2ae2ec9c19cf12ab22d57a8134a9285be' ROS_REPO=ros ABICHECK=true
````
In case of `ros_canopen` this case example this case should read:
```
- ROS_DISTRO=kinetic ABICHECK=true ABICHECK_URL=https://github.com/ros-industrial/ros_canopen/archive/0.7.5.tar.gz
```

mathias-luedtke commented on doc/index.rst:
````diff
@@ -251,6 +251,29 @@ So either `ABICHECK_URL` or `ABICHECK_FALLBACK` must be set.
+- ROS_DISTRO=indigo _EXTERNAL_REPO='ros/actionlib#38ce66e2ae2ec9c19cf12ab22d57a8134a9285be' ROS_REPO=ros ABICHECK=true
+```
+
+(TODO: Which branch and commit is going to be used for the check in this case?)
````
The branch that `industrial_ci` is run for (e.g. `kinetic-devel` of `ros_canopen`) gets tested against the `0.7.5` tag.

mathias-luedtke commented on doc/index.rst:
```diff
@@ -251,6 +251,29 @@ So either `ABICHECK_URL` or `ABICHECK_FALLBACK` must be set.
+
+(TODO: Which branch and commit is going to be used for the check in this case?)
+
+Case-2: Check ABI compatibility between the specific 2 versions.
```
This case is not supported!

mathias-luedtke commented on doc/index.rst:
```diff
@@ -251,6 +251,29 @@ So either `ABICHECK_URL` or `ABICHECK_FALLBACK` must be set.
+The version of the repo passed in `_EXTERNAL_REPO` ([ros-industrial/ros_canopen#0.7.5](https://github.com/ros-industrial/ros_canopen/releases/tag/0.7.5)) wiii be checked against the tarball of the version specified in `ABICHECK_URL`.
+
+
+
```
Case 2: Check ABI compatibility for merges (e.g. Pull request):
```
- ROS_DISTRO=kinetic ABICHECK=true ABICHECK_FALLBACK=https://github.com/ros-industrial/ros_canopen/archive/kinetic-devel.tar.gz
```
In a pull request or for merge commits the ABI gets checked for the merged version against the first parent (e.g. target branch of the pull request).
However, for push/branch test jobs the HEAD commit might not be a merge;in this case the `ABICHECK_FALLBACK` will be used instead.

130s committed:
WIP: Add catkin_lint feature.

An enhancement mentioned in https://github.com/ros-industrial/industrial_ci/issues/98

Some considerations:
- By default `catkin_lint` is run with `--strict` option, which makes the job fail even when there's only warning returned by catkin_lint. I'm not sure if we want to allow disabling --strict option on CI (those who just want the result without failing the CI job can run catkin_lint locally IMO).
- `catkin_lint` runs between build-install and tests, which I'm not sure is the best location.

130s committed:
[doc] Update for ABI checking feature.

130s commented:
Updated a commit to for the improvement based on the feedback.

I'm not yet sure if I can explain ABI_FALLBACK in a way the users can understand easily though.

130s committed:
[catkin_lint] Run against industrial_ci only.

mathias-luedtke commented:
> I'm not yet sure if I can explain ABI_FALLBACK in a way the users can understand easily though.

That's why I proposed https://github.com/ros-industrial/industrial_ci/pull/199#issuecomment-334928469, what do you think?

130s committed:
[doc][ABI check] Change ABICHECK_FALLBACK -> ABICHECK_MERGE as a more intuitive name.

130s commented:
> That's why I proposed https://github.com/ros-industrial/industrial_ci/pull/199#issuecomment-334928469, what do you think?

I missed that. That looks more friendly. I've updated the commit according to that suggestion, where I assume the order that each variable will be checked as the following chain (I didn't check if the code is written in this way besides `ABICHECK_MERGE` isn't implemented):

1. ABICHECK
2. ABICHECK_MERGE
3. ABICHECK_URL
4. ABICHECK_VERSION

130s commented:
It'd be also more efficient if `ABICHECK_URL` is set then `ABICHECK_MERGE` could be skipped. How do you think?

mathias-luedtke committed:
implemented ABICHECK_MERGE

mathias-luedtke commented:
> It'd be also more efficient if ABICHECK_URL is set then ABICHECK_MERGE could be skipped. How do you think?

`ABICHECK_URL`  is used to test pushes
`ABICHECK_MERGE` disables the default branch test in favor of testing the merge parent.

mathias-luedtke commented:
There are different cases:
1. No PR, `ABICHECK_MERGE` empty:  `ABICHECK_URL` gets used
1. No PR, `ABICHECK_MERGE` false:  `ABICHECK_URL` gets used
1. No PR, `ABICHECK_MERGE` true:  merge commit gets tested, fails if no merge **(not recommended)**
1. PR, `ABICHECK_MERGE` empty:  `ABICHECK_MERGE` will be set to true, merge will get tested **(auto-detection works for travis only)**
1. PR, `ABICHECK_MERGE` false:  `ABICHECK_URL` gets used
1. PR, `ABICHECK_MERGE` true:  merge will get tested

mathias-luedtke committed:
explicitly set ABICHECK_MERGE=false for external tests

130s committed:
[doc][abi check] Add usecase patterns.

130s commented:
Added more update.

Thanks for clarifying all the usecases in https://github.com/ipa-mdl/industrial_ci/pull/10#issuecomment-337632482. To me it took some time to understand those differences. Instead of trying to explain verbally, basically I just added your post to the doc.

mathias-luedtke commented:
Hmm, something went wrong with your rebase / merges. I will pull in the doc changes manually.

130s commented:
Hm, I remember I rebased to `origin/master` after I recently merged some PRs there. Sorry for hassle.

mathias-luedtke commented:
No worries, it was easy to achieve :)
